File: server/dto/spell/metadata/shield_formation_metadata_dto.py
from dataclasses import dataclass

from dto.base_dto import BaseDto
from dto.misc.coordinates_dto import CoordinatesDto


@dataclass
class ShieldFormationMetadataDto(BaseDto):
    # Using string keys in format "row,col" mapping to the square index
    squarePerCoordinates: dict[str, int]
    # List of squares, where each square is a list of coordinates
    squares: list[list[CoordinatesDto]]

File: client/src/dto/spell/metadata/ShieldFormationMetadataDto.ts
import { CoordinatesDto } from "../../misc/CoordinatesDto";

export interface ShieldFormationMetadataDto {
    squarePerCoordinates: Record<string, number>;
    squares: CoordinatesDto[][];
}

export function isShieldFormationMetadata(data: unknown) {
    return (
        data !== null &&
        typeof data === 'object' &&
        'squarePerCoordinates' in data &&
        'squares' in data &&
        Array.isArray(data.squares)
    );
} 